src/implementation/wasm32/simd128.rs

```diff
@@ -0,0 +1,285 @@
+            )
+            .0,
+            self.0,
+        ))
+    }
+
+    #[inline]
+    #[allow(clippy::cast_possible_wrap)]
+    unsafe fn splat(val: u8) -> Self {
+        Self::from(u8x16_splat(val))
+    }
+
+    #[inline]
+    #[allow(clippy::cast_possible_wrap)]
+    unsafe fn splat0() -> Self {
+        Self::from(u8x16(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0))
+    }
+
+    #[inline]
+    unsafe fn or(self, b: Self) -> Self {
+        Self::from(v128_or(self.0, b.0))
+    }
+
+    #[inline]
+    unsafe fn and(self, b: Self) -> Self {
+        Self::from(v128_and(self.0, b.0))
+    }
+
+    #[inline]
+    unsafe fn xor(self, b: Self) -> Self {
+        Self::from(v128_xor(self.0, b.0))
+    }
+
+    #[inline]
+    unsafe fn saturating_sub(self, b: Self) -> Self {
+        Self::from(u8x16_sub_sat(self.0, b.0))
+    }
+
+    // ugly but shr<N> requires const generics
+
+    #[allow(clippy::cast_lossless)]
+    #[inline]
+    unsafe fn shr4(self) -> Self {
+        Self::from(u8x16_shr(self.0, 4))
+    }
+
+    // ugly but prev<N> requires const generics
+
+    // TODO make this into a macro
+
+    #[allow(clippy::cast_lossless)]
+    #[inline]
+    unsafe fn prev1(self, prev: Self) -> Self {
+        Self::from(u8x16_shuffle::<
+            15,
+            16,
+            17,
+            18,
+            19,
+            20,
+            21,
+            22,
+            23,
+            24,
+            25,
+            26,
+            27,
+            28,
+            29,
+            30,
+        >(prev.0, self.0))
+    }
+
+    // ugly but prev<N> requires const generics
+
+    #[allow(clippy::cast_lossless)]
+    #[inline]
+    unsafe fn prev2(self, prev: Self) -> Self {
+        Self::from(u8x16_shuffle::<
+            14,
+            15,
+            16,
+            17,
+            18,
+            19,
+            20,
+            21,
+            22,
+            23,
+            24,
+            25,
+            26,
+            27,
+            28,
+            29,
+        >(prev.0, self.0))
+    }
+
+    // ugly but prev<N> requires const generics
+
+    #[allow(clippy::cast_lossless)]
+    #[inline]
+    unsafe fn prev3(self, prev: Self) -> Self {
+        Self::from(u8x16_shuffle::<
+            13,
+            14,
+            15,
+            16,
+            17,
+            18,
+            19,
+            20,
+            21,
+            22,
+            23,
+            24,
+            25,
+            26,
+            27,
+            28,
+        >(prev.0, self.0))
+    }
+
+    #[inline]
+    unsafe fn unsigned_gt(self, other: Self) -> Self {
+        Self::from(u8x16_gt(self.0, other.0))
+    }
+
+    #[inline]
+    unsafe fn any_bit_set(self) -> bool {
+        v128_any_true(self.0)
+    }
+
+    #[inline]
+    unsafe fn is_ascii(self) -> bool {
+        u8x16_bitmask(self.0) == 0
+    }
+}
+
+impl From<v128> for SimdU8Value {
+    fn from(v: v128) -> Self {
```

Comment:
```suggestion
    #[inline]
    fn from(v: v128) -> Self {
```

This needs to be inline, otherwise:
```
$ RUSTFLAGS="-C target-feature=+simd128" cargo asm --target=wasm32-wasi --features wasm32_simd128
<&T as core::fmt::Debug>::fmt
<() as core::fmt::Debug>::fmt
<simdutf8::basic::Utf8Error as core::fmt::Debug>::fmt
<simdutf8::basic::Utf8Error as core::fmt::Display>::fmt
<simdutf8::compat::Utf8Error as core::fmt::Debug>::fmt
<simdutf8::compat::Utf8Error as core::fmt::Display>::fmt
core::ptr::drop_in_place<&u8>
simdutf8::implementation::helpers::get_compat_error
simdutf8::implementation::wasm32::simd128::<impl core::convert::From<core::core_arch::wasm32::simd128::v128> for simdutf8::implementation::helpers::SimdU8Value<core::core_arch::wasm32::simd128::v128>>::from
simdutf8::implementation::wasm32::validate_utf8_basic_simd128
simdutf8::implementation::wasm32::validate_utf8_compat_simd128
```